lib/media/segment_prefetch.js

```diff
@@ -39,44 +40,61 @@ shaka.media.SegmentPrefetch = class {
     /** @private {shaka.media.SegmentPrefetch.FetchDispatcher} */
     this.fetchDispatcher_ = fetchDispatcher;
 
+    /** @private {boolean} */
+    this.deleteOnGet_ = deleteOnGet;
+
     /**
      * @private {!Map.<
-     *        !(shaka.media.SegmentReference|shaka.media.InitSegmentReference),
+     *        !(shaka.media.SegmentReference),
      *        !shaka.media.SegmentPrefetchOperation>}
      */
     this.segmentPrefetchMap_ = new Map();
+
+    /**
+     * @private {!Map.<
+     *        !(shaka.media.InitSegmentReference),
+     *        !shaka.media.SegmentPrefetchOperation>}
+     */
+    this.initSegmentPrefetchMap_ = new Map();
+  }
+
+  /**
+   * @return {number}
+   */
+  getLastKnownPosition() {
+    return this.prefetchPosTime_;
   }
 
   /**
-   * Fetch next segments ahead of current segment.
+   * Fetch next segments ahead of current time.
    *
-   * @param {(!shaka.media.SegmentReference)} startReference
+   * @param {number} currTime
    * @param {boolean=} skipFirst
    * @public
    */
-  prefetchSegments(startReference, skipFirst = false) {
+  prefetchSegmentsByTime(currTime, skipFirst = false) {
     goog.asserts.assert(this.prefetchLimit_ > 0,
         'SegmentPrefetch can not be used when prefetchLimit <= 0.');
 
     const logPrefix = shaka.media.SegmentPrefetch.logPrefix_(this.stream_);
     if (!this.stream_.segmentIndex) {
-      shaka.log.info(logPrefix, 'missing segmentIndex');
+      shaka.log.debug(logPrefix, 'missing segmentIndex');
       return;
     }
-    const currTime = startReference.startTime;
     const maxTime = Math.max(currTime, this.prefetchPosTime_);
     const iterator = this.stream_.segmentIndex.getIteratorForTime(maxTime);
     if (!iterator) {
       return;
     }
-    let reference = startReference;
+    let reference = iterator.next().value;
     if (skipFirst) {
       reference = iterator.next().value;
-      if (reference &&
-          reference.startTime == startReference.startTime &&
-          reference.endTime == startReference.endTime) {
-        reference = null;
-      }
+    }
+    if (!reference) {
+      return;
+    }
+    if (reference.initSegmentReference) {
```

Comment:
This if should be moved inside the while loop and take into account whether prefetchAllowed (I would move it to line 110-111)